text/3861-iter-macro.md

```diff
@@ -0,0 +1,496 @@
+assert_eq!(it.next(), Some(1));
+assert_eq!(it.next(), Some(2));
+assert_eq!(it.next(), None);
+```
+
+Note that iterator closure expressions can be directly called simply by adding parentheses after the expression, with no need to wrap the whole iterator expression in parentheses.
+In other words, `iter!(|| {})()` is valid.
+
+Within an iterator closure body, `yield` expressions are allowed.
+These have the following syntax:
+
+    YieldExpr ->
+        Expr.yield
+
+Yield expressions can be written as `foo.yield`.
+A `yield` expression always has type `()`.
+
+When evaluating a `yield`, the iterator suspends execution and passes the value of its argument to the iterating context (i.e. the caller of `next()`).
+For example, for an iterator `iter` when executing a call to `iter.next()`, if the execution of the body proceeds to a `yield 3` expression, then `iter.next()` will return `Some(3)`.
+The subsequent call to `iter.next()` will resume executing the iterator immediately after that `yield` expression.
+
+When the iterator exits the closure's body, such as by executing to the end of the body or by executing a `return` expression, the corresponding call to `next()` returns `None`, indicating the end of the iterator.
+
+The body of an iterator closure must have a return value compatible with the `()`.
+In other words, it must either return `()` or never return.
+
+Auto traits such as `Send` and `Sync` are computed separately for the iterator closure and the iterator returned by the closure.
+For iterator closures, these are computed the same as for any other closure.
+For the corresponding iterator, auto traits are computed based on the values that must be saved across a `yield` point, similar to how auto traits for an `async` block are computed based on the values that are saved across `await` points.
+
+It is not possible to hold a borrow of a value local to the iterator across a `yield` point.
+Similarly, it is not possible to `yield` a reference to an iterator-local value.
+Note that holding and yielding references that completely outlive the iterator closure (such as a reference captured from the environment) are allowed.
+
+# Drawbacks
+
+- Iterator closures have some limitations around self-borrows and lending.
+- The macro-based syntax looks less built-in than if it were actually built-in.
+- Does not exactly support the plain iterator block use case. You always have to create and invoke a closure.
+
+# Rationale and alternatives
+
+This incremental step forward will help us make continued progress towards more advanced generators.
+If we do not take this step, we expect we will see little further development in generators and users will remain frustrated that they do not have convenient syntax for creating iterators.
+This is a feature that is often missed by Rust programmers.
+
+## Why Iterator Closures?
+
+It might seem more obvious to have `iter!` evaluate directly to an `impl Iterator` with no intermediate closure step.
```

Comment:
I think the other piece I'd like elaborated here is why we need the closure-ness as part of the macro syntax too.  I was very much expecting that if this is a "first step convenience thing" that it's be just
```rust
iter! {
    1.yield;
    2.yield;
}
```
and the if you want `|| iter! { ... }` or `|item| iter! { ... }` you'd write that.  Or maybe that there'd be `into_iter! { ... }` or something as well if needed.

---

I guess the closure in the macro makes sense if you have to `()` it to get an iterator, though...